lib/connection.js

```diff
@@ -355,48 +355,46 @@ class Connection extends EventEmitter {
     });
     const rejectUnauthorized = this.config.ssl.rejectUnauthorized;
     const verifyIdentity = this.config.ssl.verifyIdentity;
-    const host = this.config.host;
+    const servername = this.config.host
 
     let secureEstablished = false;
-    const secureSocket = new Tls.TLSSocket(this.stream, {
-      rejectUnauthorized: rejectUnauthorized,
-      requestCert: true,
-      secureContext: secureContext,
-      isServer: false
+    this.stream.removeAllListeners('data');
+    const secureSocket = Tls.connect({
+      rejectUnauthorized,
+      requestCert: rejectUnauthorized,
+      secureContext,
+      isServer: false,
+      socket: this.stream,
+      servername
+    }, () => {
+      secureEstablished = true;
+      if (rejectUnauthorized) {
+        if (typeof servername === 'string' && verifyIdentity) {
+          const cert = secureSocket.getPeerCertificate(true);
+          const serverIdentityCheckError = Tls.checkServerIdentity(servername, cert);
+          if (serverIdentityCheckError) {
+            onSecure(serverIdentityCheckError);
+            return;
+          }
+        }
+      }
+      onSecure();
     });
-    if (typeof host === 'string') {
-      secureSocket.setServername(host);
-    }
     // error handler for secure socket
-    secureSocket.on('_tlsError', err => {
```

Comment:
we probably need to emit this error too or else more people will run into this issue

Reply:
I had to change event name from "connect" to "secureConnect" as well, looks like node emits both and the first one is deprecated. TLS docs only mention secureConnect - https://nodejs.org/dist/latest-v20.x/docs/api/tls.html#event-secureconnection ( later moved code to the Tls.connect() callback to avoid referencing event name )

Reply:
I might need to investigate further if this is how I should handle errors. `_tlsError` event is still in latest node - https://github.com/nodejs/node/blob/d9438ccbd8f6cbd6c8ebfae84b0f2782d2c1fca7/lib/_tls_wrap.js#L991-L1010 , not sure if documented.